Support Kerberos authentication between nodes

anusudarsan opened the pull request:
Supersedes https://github.com/prestodb/presto/pull/8859.

kokosing commented on presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:
```diff
@@ -221,6 +224,29 @@ protected void setup(Binder binder)
+            checkArgument(kerberosPrincipal != null, "kerberos principal must be set");
+            checkArgument(kerberosServiceName != null, "kerberos service name must be set");
+
+            configBinder(binder).bindConfigGlobalDefaults(KerberosConfig.class, config -> {
```
how does it relate to `KerberosConfig` configured in com/facebook/presto/server/security/ServerSecurityModule.java:50

What will happen if internal kerberos communication does not match server kerboros communication? WIll one overwrite the other?

kokosing commented on presto-main/src/main/java/com/facebook/presto/util/KerberosPrincipal.java:
```diff
@@ -0,0 +1,201 @@
+
+public class KerberosPrincipal
+{
+    private static final char REALM_SEPARATOR = '@';
```
Maybe you could have internal static class `Parser` and store parsing related code there. What do you think? Having this class defined at the bottom would not polute this file and would make the most important things to be at the top of the class.

kokosing commented on presto-main/src/main/java/com/facebook/presto/util/KerberosPrincipal.java:
```diff
@@ -0,0 +1,201 @@
+
+    private final String username;
+    private final Optional<String> hostname;
+    private final Optional<String> realm;
```
put this below static methods. `Parser` class would be helpful here, as it would make this code more readable.

kokosing commented on presto-main/src/main/java/com/facebook/presto/util/KerberosPrincipal.java:
```diff
@@ -0,0 +1,201 @@
+        builder.append(username);
+        hostname.ifPresent(hostname -> builder.append(PARTS_SEPARATOR).append(hostname));
+        realm.ifPresent(realm -> builder.append(REALM_SEPARATOR).append(realm));
+        return builder.toString();
```
If `hostname` or `realm` uses `PARTS_SEPARATOR` or `REAL_SEPARATOR` in its value, don't you need to use `ESCAPE_CHARACTER` there?

kokosing commented on presto-main/src/test/java/com/facebook/presto/util/TestKerberosPrincipal.java:
```diff
@@ -0,0 +1,114 @@
+public class TestKerberosPrincipal
+{
+    @Test
+    public void testPrincipalParsing()
```
can you also test `toString` method?

kokosing commented on presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:
```diff
@@ -242,7 +242,7 @@ protected void setup(Binder binder)
             configBinder(binder).bindConfigGlobalDefaults(HttpClientConfig.class, config -> {
                 config.setAuthenticationEnabled(true);
-                config.setKerberosPrincipal(kerberosPrincipal);
+                config.setKerberosPrincipal(principal.substituteHostnamePlaceholder().toString());
```
should not we use this here also: com/facebook/presto/server/security/KerberosAuthenticator.java:68?

kokosing commented on presto-product-tests/conf/docker/multinode-tls-kerberos/docker-compose.yml:
```diff
@@ -0,0 +1,54 @@
+        aliases:
+         - presto-master.docker.cluster
+    volumes:
+      - ../../../conf/presto/etc/environment-specific-catalogs/singlenode-kerberos-hdfs-no-impersonation/hive.properties:/docker/volumes/conf/presto/etc/catalog/hive.properties
```
why no impersonation?

kokosing commented on .travis.yml:
```diff
@@ -98,6 +98,11 @@ script:
         multinode-tls -g smoke,cli,group-by,join,tls
     fi
+  - |
+    if [[ -v PRODUCT_TESTS_SPECIFIC_ENVIRONMENT ]]; then
```
add description of this environment to `README` file

kokosing commented on presto-product-tests/conf/presto/etc/multinode-tls-kerberos-worker.properties:
```diff
@@ -0,0 +1,67 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
double licence

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestTablePartitioningInsertInto.java:
```diff
@@ -45,7 +46,7 @@ public Requirement getRequirements(Configuration configuration)
     }
 
-    @Test(groups = {HIVE_CONNECTOR, SMOKE})
+    @Test(groups = {HIVE_CONNECTOR, SMOKE, STATS_CLIENT})
```
why it fails?

kokosing commented on presto-product-tests/conf/tempto/tempto-configuration-for-docker-tls-kerberos.yaml:
```diff
@@ -0,0 +1,48 @@
+hdfs:
```
do you need to define new tempto configuration file, I thought you reuse file which is used other kerberized environment

anusudarsan commented:
@electrum I have also addressed your comments from here-
 https://github.com/prestodb/presto/pull/8859#pullrequestreview-59595130 . Please take a look.

anusudarsan commented:
related - https://github.com/prestodb/presto/pull/10317

panditsurabhi commented:
We are awaiting this change. Is it going to be merged soon? What's the ETA?

kokosing commented:
@anusudarsan Could you please rebase?
@arhimondr Could you please review?

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/InternalCommunicationModule.java:
```diff
@@ -30,5 +34,31 @@ protected void setup(Binder binder)
+            KerberosInternalCommunicationConfig internalKerberosConfig = buildConfigObject(KerberosInternalCommunicationConfig.class);
+            com.facebook.presto.server.security.KerberosConfig serverKerberosConfig = buildConfigObject(com.facebook.presto.server.security.KerberosConfig.class);
+
+            checkState(serverKerberosConfig.getKeytab() != null && serverKerberosConfig.getKeytab().getAbsolutePath().equals(internalKerberosConfig.getKerberosKeytab().getAbsolutePath()), "kerberos keytab values must match");
```
Since we are doing these checks, do we actually need 2 type of properties (for the client and for the server). Can we just use these properties from the server config?

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/KerberosInternalCommunicationConfig.java:
```diff
@@ -0,0 +1,106 @@
+
+public class KerberosInternalCommunicationConfig
+{
+    private String kerberosPrincipal;
```
You can build it based on the service name. `kerberosPrincipal=<kerberosServiceName>/_HOST`, and it should get the realm from the config. See https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/server/security/KerberosAuthenticator.java#L68

We are using this convention to build the service principal for the server. I thing we can safely follow the approach to build a client principal as well.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/KerberosInternalCommunicationConfig.java:
```diff
@@ -0,0 +1,106 @@
+public class KerberosInternalCommunicationConfig
+{
+    private String kerberosPrincipal;
+    private String kerberosServiceName;
```
Use the one from the server config, remove this property

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/KerberosInternalCommunicationConfig.java:
```diff
@@ -0,0 +1,106 @@
+{
+    private String kerberosPrincipal;
+    private String kerberosServiceName;
+    private File kerberosKeytab;
```
Use the one from the server config, remove this property

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/KerberosInternalCommunicationConfig.java:
```diff
@@ -0,0 +1,106 @@
+    private String kerberosPrincipal;
+    private String kerberosServiceName;
+    private File kerberosKeytab;
+    private File kerberosConfig;
```
Use the one from the server config, remove this property. (this one is a static one property, JVM wide)

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/KerberosInternalCommunicationConfig.java:
```diff
@@ -0,0 +1,106 @@
+    private String kerberosServiceName;
+    private File kerberosKeytab;
+    private File kerberosConfig;
+    private boolean kerberosUseCanonicalHostname = true;
```
After you remove all the other properties, this one is gonna be the only one left. I would just move it to the `InternalCommunicationConfig `

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/KerberosInternalCommunicationConfig.java:
```diff
@@ -0,0 +1,106 @@
+    private File kerberosKeytab;
+    private File kerberosConfig;
+    private boolean kerberosUseCanonicalHostname = true;
+    private File kerberosCredentialCache;
```
`credentialsCache` and `keytab` authentications are mutually exclusive. I don't think we support authentication by the `credentialsCache` (ticketCache). I would remove this

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/InternalCommunicationConfig.java:
```diff
@@ -58,4 +59,16 @@ public InternalCommunicationConfig setKeyStorePassword(String keyStorePassword)
+        return kerberosEnabled;
+    }
+
+    @Config("internal-communication.authentication.kerberos.enabled")
```
`internal-communication.kerberos.enabled`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/KerberosInternalCommunicationConfig.java:
```diff
@@ -0,0 +1,106 @@
+        return kerberosUseCanonicalHostname;
+    }
+
+    @Config("internal-communication.authentication.krb5.use-canonical-hostname")
```
`internal-communication.kerberos.use-canonical-hostname`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/InternalCommunicationModule.java:
```diff
@@ -13,6 +13,7 @@
  */
 package com.facebook.presto.server;
 
+import com.facebook.presto.util.KerberosPrincipal;
```
If you build the principal name based on the service name you will no longer need this commit. Feel free to drop it.

arhimondr commented on .travis.yml:
```diff
@@ -139,6 +139,10 @@ script:
     fi
+    if [[ -v PRODUCT_TESTS_SPECIFIC_ENVIRONMENT ]]; then
+      presto-product-tests/bin/run_on_docker.sh \
+        multinode-tls-kerberos -g smoke,cli,group-by,join,tls -x stats_client
```
Personally i would just drop the `smoke` group here, so you don't have do exclude the `stats_client` group.

arhimondr replied:
The idea with `group-by` and `join` here is to make the Presto do the exchanges.

arhimondr commented on presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos.yaml:
```diff
@@ -24,6 +24,8 @@ databases:
   presto:
     host: presto-master.docker.cluster
     port: 7778
+    http_port: 8080
```
Why is it needed? http is disabled

arhimondr commented on presto-docs/src/main/sphinx/security/internal-communication.rst:
```diff
@@ -113,6 +113,26 @@ To enable SSL/TLS for Presto internal communication, do the following:
+.. note::
+
+    The service name and keytab file used for internal Kerberos authentication is
+    taken from Coordinator Kerberos authentication properties, documented in :doc:`Kerberos</security/server>`,
```
`s/Coordinator/server`

anusudarsan commented:
@arhimondr addressed comments

anusudarsan commented:
@arhimondr ping :)

anusudarsan commented on presto-main/src/main/java/com/facebook/presto/server/InternalCommunicationModule.java:
```diff
@@ -30,5 +39,39 @@ protected void setup(Binder binder)
+        return binder -> {
+            InternalCommunicationConfig clientKerberosConfig = buildConfigObject(InternalCommunicationConfig.class);
+            com.facebook.presto.server.security.KerberosConfig serverKerberosConfig = buildConfigObject(com.facebook.presto.server.security.KerberosConfig.class);
+            verify(serverKerberosConfig.getKeytab() != null);
```
maybe also point to `internal-communication.kerberos.enabled` in the message?

arhimondr replied:
Yup, that would work
